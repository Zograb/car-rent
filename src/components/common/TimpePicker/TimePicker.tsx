import type { FC, ReactNode } from 'react';
import type { CSSProp } from 'styled-components';
import { useCallback, useState } from 'react';
import {
  type ReactDatePickerProps,
  default as RDatePicker,
} from 'react-datepicker';
import tw from 'twin.macro';
import cx from 'classnames';

// Components
import { TimePickerInput, TimePickerContainer } from './picker';

export interface TimePickerProps extends Omit<ReactDatePickerProps, 'onChange'> {
  // the date picker input will occupy the entire width if true
  fullWidth?: boolean;

  // icon component to render on input
  icon?: ReactNode;

  // custom styles
  customStyles?: CSSProp,

  // event that will be called on every date change
  onChange?: (date: Date) => void;
}

const TimePicker: FC<TimePickerProps> = ({
  customStyles = tw``,
  fullWidth = false,
  icon = '',
  onChange = () => {},
  ...otherProps
}) => {
  const [date, setDate] = useState<Date | null>(new Date());

  const handleChange = useCallback((date: Date) => {
    setDate(date);

    if (onChange) {
      onChange(date);
    }
  }, [onChange]);

  return (
    <RDatePicker
      selected={date}
      dateFormat="h:mm aa"
      css={[
        tw`h-12 rounded-lg outline-none text-gray3`,
        tw`border border-gray4 border-solid`,
        !icon && tw`pl-4`,
        icon && tw`pl-14`,
        fullWidth && tw`w-full`,
        customStyles,
      ]}
      customInput={<TimePickerInput icon={icon} fullWidth={fullWidth} />}
      popperClassName="w-full max-w-calendar"
      popperPlacement="bottom-start"
      wrapperClassName={cx({
        'w-full': fullWidth,
        'w-auto': !fullWidth,
      })}
      calendarContainer={TimePickerContainer}
      onChange={handleChange}
      showTimeSelect
      showTimeSelectOnly
      {...otherProps}
    />
  );
}

export default TimePicker;
