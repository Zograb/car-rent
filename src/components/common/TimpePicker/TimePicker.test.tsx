import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import tw from 'twin.macro';

import { TimePicker } from './TimePicker';

describe('TimePicker', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2000, 12, 12));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    const { container } = render(<TimePicker />);
    expect(container).toMatchSnapshot();
  });

  it('should render the component', () => {
    const { getByRole } = render(<TimePicker />);
    const input = getByRole('textbox');
    expect(input).toBeInTheDocument();
  });

  it('should call onChange when a time is selected', async () => {
    const user = userEvent.setup({
      advanceTimers: jest.advanceTimersByTime,
    });
    const mockOnChange = jest.fn();
    const { getByRole, findByText } = render(<TimePicker onChange={mockOnChange} />);
    const input = getByRole('textbox');
    await user.click(input);
    const timeButton = await findByText('1:00 AM');
    await user.click(timeButton);
    expect(mockOnChange).toHaveBeenCalled();
  });

  it('should render full width input', async () => {
    const { getByRole } = render(<TimePicker fullWidth />);
    const input = getByRole('textbox');
    expect(input).toHaveStyleRule('width', tw`w-full`.width as jest.Value);
  });

  it('should render time picker with custom styles', () => {
    const style = tw`bg-red-500`;
    const { getByRole } = render(<TimePicker customStyles={style} />);
    const input = getByRole('textbox');
    expect(input).toHaveStyleRule('background-color', style.backgroundColor as jest.Value);
  });
});
