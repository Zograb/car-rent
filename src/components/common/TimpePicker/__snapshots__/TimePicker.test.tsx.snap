// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker should match snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
}

.c1 {
  position: absolute;
  left: 1rem;
  top: 50%;
  --tw-translate-y: -50%;
  transform: translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  font-size: 1.5rem;
  line-height: 2rem;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
}

.c2 {
  height: 3rem;
  border-radius: 0.5rem;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
  outline: 2px solid transparent;
  outline-offset: 2px;
  border-width: 1px;
  border-style: solid;
  --tw-border-opacity: 1;
  border-color: rgb(222 222 222 / var(--tw-border-opacity));
  padding-left: 1rem;
}

<div>
  <div
    class="react-datepicker-wrapper w-auto"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="c0"
      >
        <div
          class="c1"
        />
        <input
          class="c2"
          value="12:00 AM"
        />
      </div>
    </div>
  </div>
</div>
`;
